### PR TITLE
V1 revisions

### PR DESCRIPTION
**Pull request by fadeenk**
Couple changes I would propose but would like feedback on before moving forward with them.

- Not exposing the clear & transform function
- Update the pruner function to only take in the room and not follow the currying style. Since the new scope of the function automatically inherits the id, the only other parameter needed for it is the room

Any other changes needed prior to releasing this branch. Would like to get this and [primus-rooms](https://github.com/cayasso/primus-rooms) updated asap

TODO:

- [ ] Add tests

### CODE REVIEW

**Comment by lpinca** [comment]
@fadeenk I didn't look at this and I only briefly looked at the 1.0 branch.
I think it makes sense to turn [these](https://github.com/cayasso/primus-rooms-adapter/blob/94c3a2208303f4cb4cc71f78f449a980f4996c73/src/index.js#L11-L12) into maps and each element inside them into a set.

If you already did that ignore this comment.

**Comment by fadeenk** [comment]
Do we want to do sets though? Doing so will mean we are dropping support for any native node version < 6.0, and require babel for older versions

**Comment by lpinca** [comment]
`Map` and `Set` are supported in Node.js 4.
It's a major version anyway and anything below Node.js < 4 is no longer supported so I think it makes sense. I'd prefer to not use babel though which means no destructuring, spread operator, etc.

File: src/index.js [hunk 1]
```diff
@@ -13,9 +13,9 @@ export default options => {
 
   /**
    * Get rooms socket is subscribed to.
-   *
-   * @param {String} id
-   * @param {Function} [fn]
+   * Used for get rooms
```
Comment (lpinca) [inline src/index.js]: Any reason for these lines? Isn't the above description sufficient?

File: src/index.js [hunk 2]
```diff
@@ -13,9 +13,9 @@ export default options => {
...
+   * @param {String} id - spark's ID
+   * @param {Function} [fn] - callback
```
Comment (lpinca) [inline src/index.js]: I think we no longer want the callback to be optional. Also should we switch to promises?

File: src/index.js [hunk 3]
```diff
@@ -138,81 +156,62 @@ export default options => {
...
+     * Helper for broadcast
+     * @param {String[]} ids - spark IDs to send to
+     * @param {Object} clients - connected clients
+     * @param {*} data - the data to send
+     * @param {Object} options - same options as broadcast
+     * @api private
+     */
+
+    function send(ids, clients, data, options) {
+        const { except = [], method = 'writer', transformer } = options
```
Comment (lpinca) [inline src/index.js]: Typo on `method` default value, should be `'write'`.

**Comment by lpinca** [comment]
> Not exposing the clear & transform function

I'm :+1: if the functions are private and not needed for testing.

> Update the pruner function to only take in the room and not follow the currying style. Since the new scope of the function automatically inherits the id, the only other parameter needed for it is the room

Can you please show an example? If I understand correctly this will not work.

File: src/index.js (hunk 1 above)
Comment (fadeenk) [reply src/index.js]: It is sufficient, however following the [JSDoc](http://usejsdoc.org/tags-param.html) specs for description to ensure it is clearer/easier for people to jump in and understand what is going on

File: src/index.js (hunk 2 above)
Comment (fadeenk) [reply src/index.js]: I had the same issue regarding the promises support for older node versions, so didnt do it. But since we are dropping the older version will update it to use promises :)

File: src/index.js (hunk 3 above)
Comment (fadeenk) [reply src/index.js]: Agree, it did look weird when I saw it. I wasn't sure if it was a new method added in a newer version or a typo. I didnt get the chance to investigate it, it was on this branch when i started working on it

**Comment by cayasso** [comment]
Thanks guys for collaborating!!!

File: src/index.js (hunk 1 above)
Comment (cayasso) [reply src/index.js]: Nahh!!! I would leave it as simple as possible, we don't even use JSDoc I had comments just for the sake of some clarity.

Comment (fadeenk) [reply src/index.js]: But why not just document everything clearly, so when using IDEs it picks up the documentation and make the development process much easier?

File: src/index.js (hunk 2 above)
Comment (cayasso) [reply src/index.js]: Agree!!! yep no callbacks just promise, this code was way before deciding on promises

File: src/index.js (hunk 3 above)
Comment (cayasso) [reply src/index.js]: Upss sorry my fault!!!

File: src/index.js (hunk 1 above)
Comment (cayasso) [reply src/index.js]: I am totally with u, its just that two lines of comments look redundant, I agree with extending the comment above that line. ;-)

**Comment by cayasso** [comment]
> Not exposing the clear & transform function

@fadeenk `clear` and `transform` are public methods, if you check closely `clear` and `empty` serve different purposes and are called individually by `primus-rooms`.

https://github.com/cayasso/primus-rooms/blob/master/lib/rooms.js#L263
https://github.com/cayasso/primus-rooms/blob/master/lib/rooms.js#L264

> Update the pruner function to only take in the room and not follow the currying style. Since the new scope of the function automatically inherits the id, the only other parameter needed for it is the room

I am up with the idea of putting pruner in `del` this will eliminate the extra function call, however if we are looking for performance with this change then an even better way is to remain pruner outside of `del` remove the curry style but make it receive the `id` and the `room`, this will avoid always creating the function when `del` is called. But this [line](https://github.com/cayasso/primus-rooms-adapter/blob/1.0/src/index.js#L59) would need to be changed.

```js
function prune(id, room) {
  delete sids[id][room]
  if (!keys(sids[id]).length) delete sids[id]
  delete rooms[room][id]
  if (!keys(rooms[room]).length) {
    delete rooms[room]
    wild.remove(room)
  }
}
```

**Comment by fadeenk** [comment]
For the clear logic I dont see why it cannot be handled within the adapter rather than `primus-rooms`. My idea was the plugin will just call the `empty` method with rooms and if it is undefined then the adapter will internally call clear. Less exposed APIs = less confusion on the user's end.

I agree regarding the prune logic. I am updating get/add/del/prune to use promises as well as adding tests should be ready soon and would love your feedback on those.

File: src/index.js [hunk 4]
```diff
@@ -213,6 +223,37 @@ export default options => {
...
+          socketsMap.delete(id);
+        }
+      }
+      return Promise.resolve({success: true});
+    } catch (e) {
+      return Promise.reject(e);
+    }
+  }
+  
+  return { getRooms, add, del, clients, broadcast, empty }
```
Comment (cayasso) [inline src/index.js]: Why the change to the `get` method name?

**Comment by fadeenk** [comment]
@cayasso  @lpinca I have couple questions

What should the returns for the promises be?
For `get` now `getRooms`, do we want to return an Array or a Set?
Are we going support wildcard plugin, if so are we looking to keep it as is or use `Set` instead of `Array` or do you have some other changes in mind?
How do we want to approach testing, any specific structure or requirements?
Are we going to set up coveralls on this repo? Would like to have everything covered

File: src/index.js (hunk 4 above)
Comment (fadeenk) [reply src/index.js]: it makes easier to understand its function, same thing for clients would rename it to getClients. But can revert it to the way it was. Since it is a major refactor, I would update the naming to be clearer but if we dont want to do that, it is fine as is.

Comment (cayasso) [reply src/index.js]: Yes please change it back, I prefer short names over long/verbose names, the same apply for the `Set` variable names and any additional variable.

Comment (fadeenk) [reply src/index.js]: got it will change it back in a bit

**Comment by fadeenk** [comment]
All functions in the src have been rewritten, I have not touched wildcard at all.

**Comment by lpinca** [comment]
Can we remove babel?

File: src/index.js [hunk 5]
```diff
@@ -1,174 +1,231 @@
...
-    setImmediate(fn, null, rms)
+  function get(id) {
+    if (id) {
+      return Promise.resolve(Array.from(sids.get(id)));
+    }
+    let roomsArray = [];
+    rooms.forEach((value, key) => {
+      roomsArray.push(key);
+    });
+    return Promise.resolve(roomsArray);
```
Comment (lpinca) [inline src/index.js]: This can be simplified like this:

```js
return Promise.resolve(Array.from(rooms.keys()));
```

File: src/index.js [hunk 6]
```diff
@@ -1,174 +1,231 @@
...
   }
 
   /**
    * Add a socket to a room.
-   *
-   * @param {String} id
-   * @param {String} room
-   * @param {Function} [fn]
+   * Used for joining rooms
```
Comment (lpinca) [inline src/index.js]: I still find these redundant. "Add a socket to a room" - "Used for joining rooms" yeah if it adds a socket to room it is obviously used for joining rooms :)

File: src/index.js [hunk 7]
```diff
@@ -1,174 +1,231 @@
...
 
-  function add(id, room, fn = noop) {
-    sids[id] = sids[id] || []
-    sids[id][room] = true
-    rooms[room] = rooms[room] || {}
-    rooms[room][id] = true
-    wild.add(room)
-    setImmediate(fn, null, null)
+  function add(id, room) {
+    let socketRooms = sids.get(id);
```
Comment (lpinca) [inline src/index.js]: const

File: src/index.js [hunk 8]
```diff
@@ -1,174 +1,231 @@
...
+    let socketsInRoom = rooms.get(room);
+    // if socket has a rooms set just add the room
+    if (socketRooms instanceof Set) {
```
Comment (lpinca) [inline src/index.js]: I think there is not need to check if it's a Set.

File: src/index.js [hunk 9]
```diff
@@ -1,174 +1,231 @@
...
+    if (socketsInRoom instanceof Set) {
+      socketsInRoom.add(id);
+    } else {
+      // room has no set of sockets, create and add it
+      rooms.set(room, new Set([id]));
+    }
+    // add to wild rooms
+    // TODO visit wild logic to work Map and Set
+    // wild.add(room);
+    return Promise.resolve({success: true});
```
Comment (lpinca) [inline src/index.js]: Hmm not really pleased with the resolve value but I'm also not sure what to return. How about nothing (`undefined`)?

File: src/index.js [hunk 10]
```diff
@@ -1,174 +1,231 @@
...
-        wild.find(room, rms, prune)
-      } else if (sids[id] && sids[id][room]) {
-        prune(room)
-      }
-    } else {
-      for (room in sids[id]) prune(room)
+      return prune(id, room);
     }
-    setImmediate(fn, null, null)
+    let promises = [];
```
Comment (lpinca) [inline src/index.js]: const

File: src/index.js [hunk 11]
```diff
@@ -1,174 +1,231 @@
...
    */
 
-  function empty(room, fn = noop) {
-    if (room && rooms[room]) {
-      if (!isArray(room)) clear(room)
-      else for (const rm of room) clear(rm)
+  function empty(room) {
+    if (isArray(room)) {
+      let promises = [];
+      for (let rm in rooms) promises.push(clear(rm));
```
Comment (lpinca) [inline src/index.js]: I think there is a typo. If I'm not wrong `rooms` should be `room`. Also how about using `Array.prototype.map()`?

File: src/index.js [hunk 12]
```diff
@@ -1,174 +1,231 @@
...
    */
-
-  function clear(room, fn = noop) {
-    if ('function' === typeof room) {
-      rooms = {}, sids = {}
-    }
-    for (const id in rooms[room]) {
-      delete sids[id][room]
+  function clear(room) {
+    try {
```
Comment (lpinca) [inline src/index.js]: Why the `try...catch`? Is there anything that can throw here?

File: src/index.js [hunk 13]
```diff
@@ -1,174 +1,231 @@
...
-      wild.match(room, key =>
-        send(rooms[key], clients, data, options)
+      let ids = new Set();
+      sids.forEach((value, key) => {
+        ids.add(key);
+      });
+      return send(ids, clients, data, options)
     }
+    let promises = [];
+    rooms.forEach((value, room) => {
```
Comment (lpinca) [inline src/index.js]: This doesn't look correct to me. We should only loop trough the rooms that have been passed via `options.rooms`.

File: src/index.js [hunk 14]
```diff
@@ -1,174 +1,231 @@
...
-    const { except = [], method = 'writer', transformer } = options
-    let sent = {}
-    for (let id in ids) {
-      const socket = clients[id]
-      if (sent[id] || ~except.indexOf(id) || !socket) continue
-      transform(socket, data, method, transformer)
-      sent[id] = true
+    const { except = [], method = 'write', transformer } = options
+    let promises = [];
+    try {
```
Comment (lpinca) [inline src/index.js]: I think this `try...catch` block is not needed. Also we should keep the `sent` object, it probably makes sense to turn it into a set.

File: src/index.js [hunk 15]
```diff
@@ -1,174 +1,231 @@
...
 const isArray = Array.isArray
 const keys = Object.keys
 
-export default options => {
+export default (options = {}) => {
   const wild = createWildcard({ enabled: options.wildcard })
 
-  let rooms = {}
-  let sids = {}
+  let rooms = new Map();
```
Comment (lpinca) [inline src/index.js]: const

File: test/test.js [hunk 16]
```diff
@@ -17,12 +24,47 @@ describe('primus-rooms-adapter', function () {
...
     it('should not leak memory', function () {
-      var adapter = new Adapter();
       adapter.del('foo');
       expect(adapter.rooms).to.eql({});
       expect(adapter.sids).to.eql({});
     });
   });
+  
+  describe('Adapter#add', function() {
+    it('should join a new room', function(done) {
```
Comment (lpinca) [inline test/test.js]: Mocha has built-in support for promises so we can remove `done` here and the `catch` from the chain below.

File: test/test.js [hunk 17]
```diff
@@ -17,12 +24,47 @@ describe('primus-rooms-adapter', function () {
...
+      rooms.forEach((room) => {
+        promises.push(adapter.add(sparkID, room));
+      });
+      Promise.all(promises).then((responses) => {
+        responses.forEach((res) => {
+          expect(res.success).to.eql(true);
+        });
+        return adapter.get(sparkID);
+      }).then((roomsJoined) => {
+        expect(roomsJoined).to.be.instanceOf(Array);
```
Comment (lpinca) [inline test/test.js]: How about using just `expect(roomsJoined.sort()).to.deep.equal(rooms)`?

File: src/index.js (hunk 6 above)
Comment (fadeenk) [reply src/index.js]: I will be updating all the docs and cleaning them once the good is all good :)

File: src/index.js (hunk 8 above)
Comment (fadeenk) [reply src/index.js]: what if it is `undefined` the room doesn't exist??

File: src/index.js (hunk 9 above)
Comment (fadeenk) [reply src/index.js]: Same here, can change it to just resolve

File: src/index.js (hunk 12 above)
Comment (fadeenk) [reply src/index.js]: I think it might have been a left over

File: src/index.js (hunk 13 above)
Comment (fadeenk) [reply src/index.js]: you are correct that should be `rms`

File: src/index.js (hunk 14 above)
Comment (fadeenk) [reply src/index.js]: Ok, can be done

**Comment by fadeenk** [comment]
I will be less active on this PR till the weekend work and life taking most of my time

File: src/index.js (hunk 8 above)
Comment (lpinca) [reply src/index.js]: Yeah, `sids.get(id)` will return either the set or `undefined`. If return value is `undefined` it means that the socket didn't join any room.

File: src/index.js (hunk 14 above)
Comment (fadeenk) [reply src/index.js]: Ok for this one i remember now why i have the `try...catch` it is there in case the transformer function throws due to custom logic passed in

File: src/index.js [hunk 18]
```diff
@@ -1,174 +1,226 @@
...
-    } else {
-      for (room in sids[id]) prune(room)
+      return prune(id, room);
     }
-    setImmediate(fn, null, null)
+    const promises = [];
+    rooms.forEach((value, key) => {
+      promises.push(prune(id, key));
+    });
+    return Promise.all(promises);
```
Comment (lpinca) [inline src/index.js]: We should add a `.then(noop)` to ensure that the promise will resolve to `undefined` and not an empty array.

File: src/index.js [hunk 19]
```diff
@@ -1,174 +1,226 @@
...
    */
 
-  function empty(room, fn = noop) {
-    if (room && rooms[room]) {
-      if (!isArray(room)) clear(room)
-      else for (const rm of room) clear(rm)
+  function empty(room) {
+    if (isArray(room)) {
+      room.map(rm => clear(rm));
+      return Promise.all(room);
```
Comment (lpinca) [inline src/index.js]: Same here.

File: src/index.js [hunk 20]
```diff
@@ -1,174 +1,226 @@
...
-    for (const room of rms) {
-      let ids = rooms[room]
-      if (ids) send(ids, clients, data, options)
-      wild.match(room, key =>
-        send(rooms[key], clients, data, options)
+      const ids = new Set();
+      sids.forEach((value, key) => {
+        ids.add(key);
+      });
+      return send(ids, clients, data, options).then(result => new Set(result));
```
Comment (lpinca) [inline src/index.js]: Can't we use the iterator directly here? Instead of computing `ids` I think we can pass `sids.keys()` directly.

File: src/index.js [hunk 21]
```diff
@@ -1,174 +1,226 @@
...
+   * @param {String[]} options.rooms - list of the rooms to broadcast to
+   * @param {String} options.method - 'write' or 'send' if primus-emitter is present
+   * @param {Function} options.transformer - the transformer function to call on the data
+   * @param {Object} clients -  Connected clients
+   * @return {Promise.<Set>} set of sparkIDs transmitted to
    * @api public
    */
-
+  
   function broadcast(data, options = {}, clients) {
```
Comment (lpinca) [inline src/index.js]: There is a problem in this function.
The original version used the `sent` object to prevent data from being sent to same socket more than once. This can happens for example when a socket joined more rooms and we are broadcasting to those specific rooms.

The socket should receive the message only once.

File: src/index.js [hunk 22]
```diff
@@ -177,43 +229,50 @@ export default options => {
       return socket[method].apply(socket, data)
     }
 
-    let packet = {}
+    let packet = {};
```
Comment (lpinca) [inline src/index.js]: const

File: src/index.js [hunk 23]
```diff
@@ -177,43 +229,50 @@ export default options => {
...
-  function pruner(id) {
-    return room => {
-      delete sids[id][room]
-      if (!keys(sids[id]).length) delete sids[id]
-      delete rooms[room][id]
-      if (!keys(rooms[room]).length) {
-        delete rooms[room]
-        wild.remove(room)
+  function prune(id, roomName) {
+    let socketsSet = rooms.get(roomName);
```
Comment (lpinca) [inline src/index.js]: const

File: src/index.js [hunk 24]
```diff
@@ -177,43 +229,50 @@ export default options => {
...
+    // remove the socket from the socketsInRoom set
+    if (socketsSet) {
+      socketsSet.delete(id);
+      // if no sockets connect to a room remove the room
+      if (socketsSet.size === 0) {
+        rooms.delete(roomName);
+        // TODO add wild.remove and update its logic
       }
     }
+    let roomsSet = sids.get(id);
```
Comment (lpinca) [inline src/index.js]: const

File: src/index.js [hunk 25]
```diff
@@ -177,43 +229,50 @@ export default options => {
...
+    // remove room from the socket's rooms
+    if (roomsSet instanceof Set) {
```
Comment (lpinca) [inline src/index.js]: We don't have to check that it's an instance of `Set`.

File: test/test.js [hunk 26]
```diff
@@ -17,12 +24,37 @@ describe('primus-rooms-adapter', function () {
...
     });
   });
+  
+  describe('Adapter#add', function() {
+    it('should join a new room', function() {
+      let room = 'testRoom';
+      adapter.add(sparkID, room).then((res) => {
+        return adapter.get(sparkID);
+      }).then((res) => {
+        expect(res).to.be.instanceof(Array);
```
Comment (lpinca) [inline test/test.js]: `expect(res).to.deep.equal([room])`

File: test/test.js [hunk 27]
```diff
@@ -17,12 +24,37 @@ describe('primus-rooms-adapter', function () {
...
-      var adapter = new Adapter();
       adapter.del('foo');
       expect(adapter.rooms).to.eql({});
       expect(adapter.sids).to.eql({});
     });
   });
+  
+  describe('Adapter#add', function() {
+    it('should join a new room', function() {
+      let room = 'testRoom';
```
Comment (lpinca) [inline test/test.js]: const

File: test/test.js [hunk 28]
```diff
@@ -17,12 +24,37 @@ describe('primus-rooms-adapter', function () {
...
+      }).then((res) => {
+        expect(res).to.be.instanceof(Array);
+        expect(res.length).to.eql(1);
+        expect(res[0]).to.eql(room);
+      });
+    });
+    
+    it('should join multiple rooms', function() {
+      const rooms = ['test1', 'test2', 'test3'];
+      let promises = [];
```
Comment (lpinca) [inline test/test.js]: I'd use `const promises = rooms.map((room) => adapter.add(sparkID, room))`;

File: src/index.js (hunk 20 above)
Comment (fadeenk) [reply src/index.js]: the send function expects a set however we could do something like this `new Set([...sids.keys()])`

Comment (lpinca) [reply src/index.js]: Well `send()` uses `for...of` so an iterator should work as expected.

File: src/index.js (hunk 21 above)
Comment (fadeenk) [reply src/index.js]: I noticed that but forgot to update the code. The set takes care of duplicates, just need to update the logic so all the rooms' sockets are aggregated into one set then send that set

Comment (lpinca) [reply src/index.js]: Gotcha.

File: test/test.js (hunk 28 above)
Comment (fadeenk) [reply test/test.js]: i use the rooms array to check if the rooms joined are the same rooms as the ones we called the join on

File: src/index.js (hunk 20 above)
Comment (fadeenk) [reply src/index.js]: good catch

File: test/test.js (hunk 28 above)
Comment (lpinca) [reply test/test.js]: Yes, my suggestion is to replace `Array.prototype.forEach()` with `Array.prototype.map()`.

Comment (fadeenk) [reply test/test.js]: I am not seeing it, if we map the rooms => promises then we wont be able to do the deep equal on line [56](https://github.com/cayasso/primus-rooms-adapter/pull/6/commits/c5db9df0ee1497d3592cd8cfdb9da562c84b3c82#diff-c1129c8b045390789fa8ff62f2c6b4a9R56)

Comment (lpinca) [reply test/test.js]: Why not, `map` does not mutate the original array, it returns a new one.

```js
const rooms = ['test1', 'test2', 'test3'];
const promises = rooms.map((room) => adapter.add(sparkID, room));
```

`rooms` is still the same.

Comment (fadeenk) [reply test/test.js]: Sorry it is too early right now and I stayed up too late yesterday. You are correct, will fix all the changes needed when I get back from work tonight

**Comment by fadeenk** [comment]
I have been swamped with work and other things, havent had the chance to fix anything yet.
In the mean time regarding babel, if we want to use `import`, the only way I know to get it to work is via using babel. We could go back to require and everything will still work

**Comment by lpinca** [comment]
Yes, not only that, I saw that`for...of`, destructuring and other ES6 feature are also used. Those are not available in Node.js 4 and if we want to use them we have to transpile.

**Comment by fadeenk** [comment]
So how do we want to handle this, do we change the code so it works on native node 4 or maybe just support node 6 or rely on babel?

**Comment by lpinca** [comment]
I have no strong opinions, we can make the code work on Node.js 4 without babel or use a preset that will only transpile features not available on Node.js 4.

Either way works for me.

**Comment by cayasso** [comment]
I would say lets go with babel, my plan is to support `async/await` in `primus-rooms` any way so this is a good start. But lets use it wisely.

**Comment by cayasso** [comment]
I just did a refactor to the existing code and added a bunch of changes for the benefit of the project.
- Added `async/await` support without babel using just `async-to-gen`
- With the change above there is no need to include any reference to `Promise`.
- Added `XO` for code style standard
- Replace `import` with `require`
- Added `yarn.lock`
- Don't expose `clear` and `transform` as suggested here.
- Added `should` instead of `expect` for test assertion.
- And did some code cleanup and simplification.

I am planing to remove all comments from the code, I think code should explain it self.
Happy to hear your comments, the code is not tested yet because I haven't had the chance to, so there may be some errors, etc, I will take a closer look tomorrow.

**Comment by fadeenk** [comment]
@cayasso Sounds great, I will take a look tomorrow. as far as for the comments I like to keep at least JSDoc in there since it can be picked up by IDEs and some editors and makes the dev process a bit faster

**Comment by lpinca** [comment]
@cayasso destructuring assignment and default parameters are not supported in Node.js 4 so babel is still required, unless we remove those features.

**Comment by fadeenk** [comment]
I did some minor tweaks based on the new revisions done by @cayasso

File: src/index.js [hunk 29]
```diff
@@ -78,9 +78,8 @@ module.exports = exports = options => {
    */
 
   async function empty(room) {
-    if (!rooms.get(room)) return
-    if (!Array.isArray(room)) clear(room)
-    room.forEach(clear)
+    if (Array.isArray(room)) room.forEach(clear)
+    if (rooms.get(room)) clear(room)
```
Comment (lpinca) [inline src/index.js]: I would use `rooms.has(room)` and use `else if` so it is not checked if `room` is an array.

File: src/index.js [hunk 30]
```diff
@@ -92,8 +91,8 @@ module.exports = exports = options => {
 
   function clear(room) {
     const rsocks = rooms.get(room)
-    rsocks.forEach(id => socks.get(id).clear())
-    rsocks.clear()
+    rsocks.forEach(id => socks.get(id).delete(room))
```
Comment (lpinca) [inline src/index.js]: We should also remove the set from the map if it is empty.

File: src/index.js [hunk 31]
```diff
@@ -122,7 +121,7 @@ module.exports = exports = options => {
           send(rooms.get(key), clients, data, options))
       })
     } else {
-      send(socks.keys(), clients, data, options)
+      send(new Set([...socks.keys()]), clients, data, options)
```
Comment (lpinca) [inline src/index.js]: This doesn't work on Node.js 4.

Comment (fadeenk) [reply src/index.js]: according to [node.green](http://node.green/), it will work if harmony flag is set

Comment (lpinca) [reply src/index.js]: Using `--harmony` is not recommended for production environments and I wouldn't expect users to use the flag anyway :)

**Comment by fadeenk** [comment]
@lpinca @cayasso I made some updates for the changes here, if everything is good I will move on to testing

**Comment by lpinca** [comment]
@fadeenk I think that the logic to prevent a single message from being sent multiple times to the same socket is still missing.

**Comment by cayasso** [comment]
@lpinca correct me if I am wrong, `Set`s doesn't allow duplicates so there will ever be only one unique id when iterating and sending, I think this should avoid sending multiple times the same message.
https://github.com/cayasso/primus-rooms-adapter/blob/1.0/src/index.js#L125

But of course this need to be tested  ;-)

**Comment by lpinca** [comment]
@cayasso assume that a socket joined rooms 'a' and 'b' and you broadcast to room 'a' and 'b', so you first get the sockets (a set) of room 'a' and broadcast to them and then the sockets (another set) of room 'b' and also broadcast to them. The result is that the original socket received the message twice.

I can't see anything in the current [code](https://github.com/cayasso/primus-rooms-adapter/blob/b65b9bb62cd561767f74b4301095b1154ac82a23/src/index.js#L117-L124) which prevents this.

**Comment by cayasso** [comment]
I see what you mean now, I was looking only at that line I sent :(

**Comment by lpinca** [comment]
I also think there is a minor bug [here](https://github.com/cayasso/primus-rooms-adapter/blob/b65b9bb62cd561767f74b4301095b1154ac82a23/src/index.js#L117). It should be `if (rms.length) ...` if I'm not wrong.

File: src/index.js [hunk 32]
```diff
@@ -117,13 +117,15 @@ module.exports = exports = options => {
   async function broadcast(data, options, clients) {
     options = options || {}
     const rms = options.rooms || []
-    if (rms.length === 0) {
+    let ids = []
```
Comment (lpinca) [inline src/index.js]: I was thinking at a different approach. Instead of creating an array with all the ids an then removing the duplicates, use a set and then pass that set to the `send()` function.
In `send()` use the set to see if the message has already been sent to the socket.

So basically replacing this line with `const sent = new Set();` and in `send()`:

```js
function send(sent, ids, clients, data, options) {
  ...
  ids.forEach(id => {
    const socket = clients[id]
    if (!socket || ~except.indexOf(id) || sent.has(id)) return
    sent.add(id)
    transform(socket, data, method, transformer)
  })
}
```

Makes sense? Thoughts?

Comment (fadeenk) [reply src/index.js]: I like the idea and we can return the sent set as well. I can implement this approach tomorrow when I get the chance

Comment (fadeenk) [reply src/index.js]: Who needs sleep, I just updated it :)

File: src/index.js [hunk 33]
```diff
@@ -114,12 +117,13 @@ module.exports = exports = options => {
...
       rms.forEach(room => {
         const ids = rooms.get(room)
-        if (ids) send(ids, clients, data, options)
+        if (ids) send(sent, ids, clients, data, options)
         wild.match(room, key =>
-          send(rooms.get(key), clients, data, options))
+          send(sent, rooms.get(key), clients, data, options))
       })
     } else {
       send(toArray(socks.keys()), clients, data, options)
```
Comment (lpinca) [inline src/index.js]: The first argument should be `sent` here as well.

**Review by lpinca — APPROVED** [review]
